File: genetic-packer/src/main/java/genetic/packer/statistics/impl/GenerationStatisticsCreatorImpl.java
package genetic.packer.statistics.impl;

import java.util.function.Function;

import genetic.api.fitness.Fitness;
import genetic.packer.evolution.generation.dto.Generation;
import genetic.packer.evolution.generation.dto.GenerationStatistics;
import genetic.packer.evolution.generation.dto.GenerationStatisticsBuilder;
import genetic.packer.statistics.GenerationStatisticsCreator;
import javafx.scene.shape.Box;
import javaslang.collection.Seq;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

/**
 * @author piotr.larysz
 */
@Component
public class GenerationStatisticsCreatorImpl implements GenerationStatisticsCreator {

    @Autowired
    private Function<Generation<Double, Box>, Fitness<Double>> bestFitnessResolver;

    @Autowired
    private Function<Generation<Double, Box>, Double> averageFitnessResolver;

    @Override
    public Seq<GenerationStatistics> create(Seq<Generation<Double, Box>> generations) {
        return generations.map(this::createGenerationStatistics);
    }

    private GenerationStatistics createGenerationStatistics(Generation<Double, Box> generation) {
        final Fitness<Double> bestFitness = this.bestFitnessResolver.apply(generation);
        return new GenerationStatisticsBuilder()
            .bestFitness(bestFitness.get())
            .averageFitness(this.averageFitnessResolver.apply(generation))
            .bestFitnessExplained(bestFitness.explain())
            .build();
    }
}

File: genetic-statistics/src/main/java/genetic/statistics/impl/GenerationStatisticsCreatorImpl.kt
package genetic.statistics.impl

import genetic.api.fitness.Fitness
import genetic.api.generation.Generation
import genetic.api.statistics.GenerationStatistics
import genetic.statistics.GenerationStatisticsCreator

class GenerationStatisticsCreatorImpl(
    private val bestFitnessResolver: (Generation<*>) -> Fitness,
    private val averageFitnessResolver: (Generation<*>) -> Double
) : GenerationStatisticsCreator {

  override fun create(generations: Sequence<Generation<*>>) = generations.map {
    createGenerationStatistics(it)
  }

  private fun createGenerationStatistics(generation: Generation<*>): GenerationStatistics {
    return bestFitnessResolver(generation).let {
      GenerationStatistics(
          bestFitness = it.score,
          averageFitness = averageFitnessResolver(generation),
          bestFitnessExplained = it.explain(),
          averageNaturalFitness = 0.0, //FIXME
          generationNumber = generation.number
      )
    }
  }
}